File: client/src/app/modules/miscellaneous/services/notification.service.ts
import { HttpClient, HttpParams } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { MatDialog } from '@angular/material/dialog';
import { Observable, Subject } from 'rxjs';
import { AppRestService } from 'src/app/services/appRest.service';
import { Notification } from '../../miscellaneous/model/Notification';

@Injectable({
  providedIn: 'root'
})
export class NotificationService extends AppRestService<Notification> {

  notifications: Observable<number> | undefined;
  notificationsResult: number = 0;

  displayFuture: boolean = false;
  displayRead: boolean = false;

  constructor(http: HttpClient,
    public notificationDialog: MatDialog) {
    super(http, "miscellaneous");
    /*this.notifications = timer(1, NOTIFICATION_REFRESH_INTERVAL).pipe(
      switchMap(() => this.getNotificationsNumber(this.displayFuture, this.displayRead)),
      retry(),
      tap((value) => {
        this.notificationsResult = value;
        if (!this.notificationsResult)
          this.notificationsResult = 0;
      }),
      share(),
      takeUntil(this.stopPolling)
    );
    this.notifications.subscribe();*/
  }

  ngOnDestroy() {
    this.stopPolling.next(false);
  }

  private stopPolling = new Subject();

  getNotificationsNumber(displayFuture: boolean, displayRead: boolean) {
    return this.get(new HttpParams().set("displayFuture", displayFuture).set("displayRead", displayRead), "notifications/nbr") as any as Observable<number>;
  }

  refreshNotificationsNumber() {
    this.getNotificationsNumber(this.displayFuture, this.displayRead).subscribe(response => {
      this.notificationsResult = response;
      if (!this.notificationsResult)
        this.notificationsResult = 0;
    })
  }

  getNotifications(displayFuture: boolean, notificationTypes: string[], displayRead: boolean) {
    let params = new HttpParams().set("displayFuture", displayFuture).set("displayRead", displayRead);
    if (notificationTypes)
      params = params.set("notificationTypes", notificationTypes.join(","));
    return this.getList(params, "notifications");
  }

  getNotificationsResult(): number {
    if (!this.notificationsResult)
      return 0;
    return this.notificationsResult;
  }

  getIsDisplayFuture() {
    return this.displayFuture;
  }

  getAllNotificationTypes() {
    return this.getListCached(new HttpParams(), "notifications/types") as any as Observable<string[]>;
  }

  addOrUpdateNotification(notification: Notification) {
    return this.addOrUpdate(new HttpParams(), "notification", notification);
  }

  deleteNotification(notification: Notification) {
    return this.delete(new HttpParams().set("notificationId", notification.id), "notification");
  }

  addPersonnalNotification(notification: Notification) {
    return this.addOrUpdate(new HttpParams(), "notification/personnal", notification, "Notification ajoutée", "Erreur lors de l'enregistrement de la notification");
  }

  getNotificationsForCustomerOrder(customerOrderId: number) {
    return this.getList(new HttpParams().set("customerOrderId", customerOrderId), "notifications/customer-order");
  }

  getNotificationsForQuotation(quotationId: number) {
    return this.getList(new HttpParams().set("quotationId", quotationId), "notifications/quotation");
  }

  getNotificationsForProvision(provisionId: number) {
    return this.getList(new HttpParams().set("provisionId", provisionId), "notifications/provision");
  }

  getNotificationsForInvoice(invoiceId: number) {
    return this.getList(new HttpParams().set("invoiceId", invoiceId), "notifications/invoice");
  }

  getNotificationsForService(serviceId: number) {
    return this.getList(new HttpParams().set("serviceId", serviceId), "notifications/service");
  }

  getNotificationsForAffaire(affaireId: number) {
    return this.getList(new HttpParams().set("affaireId", affaireId), "notifications/affaire");
  }

  getNotificationsForTiers(tiersId: number) {
    return this.getList(new HttpParams().set("tiersId", tiersId), "notifications/tiers");
  }

  getNotificationsForResponsable(responsableId: number) {
    return this.getList(new HttpParams().set("responsableId", responsableId), "notifications/responsable");
  }

}

File: src/main/java/com/jss/osiris/modules/osiris/miscellaneous/service/NotificationService.java
package com.jss.osiris.modules.osiris.miscellaneous.service;

import java.util.List;

import com.jss.osiris.libs.exception.OsirisException;
import com.jss.osiris.modules.osiris.crm.model.Candidacy;
import com.jss.osiris.modules.osiris.miscellaneous.model.Attachment;
import com.jss.osiris.modules.osiris.miscellaneous.model.Notification;
import com.jss.osiris.modules.osiris.quotation.model.BaloNotice;
import com.jss.osiris.modules.osiris.quotation.model.BodaccNotice;
import com.jss.osiris.modules.osiris.quotation.model.CustomerOrder;
import com.jss.osiris.modules.osiris.quotation.model.JoNotice;
import com.jss.osiris.modules.osiris.quotation.model.Provision;
import com.jss.osiris.modules.osiris.quotation.model.Service;
import com.jss.osiris.modules.osiris.reporting.model.IncidentReport;

public interface NotificationService {
        public List<Notification> getNotificationsForCurrentEmployee(Boolean displayFuture, Boolean displayRead,
                        List<String> notificationTypes, Boolean onlyForNumber, Boolean completeAdditionnalInformation)
                        throws OsirisException;

        public Notification getNotification(Integer id);

        public Notification addOrUpdateNotificationFromUser(Notification notification);

        public Notification addOrUpdateNotification(Notification notification);

        public void deleteNotification(Notification notification);

        public void notifyAttachmentAddToProvision(Provision provision, Attachment attachment) throws OsirisException;

        public void notifyAttachmentAddToService(Service service, Attachment attachment)
                        throws OsirisException;

        public void notifyInformationAddToService(Service service) throws OsirisException;

        public void notifyAttachmentAddToCustomerOrder(CustomerOrder order, Attachment attachment)
                        throws OsirisException;

        public void notifyCommentFromMyJssAddToCustomerOrder(CustomerOrder order)
                        throws OsirisException;

        public void notifyGuichetUniqueFormaliteStatusValidated(Provision provision) throws OsirisException;

        public void notifyGuichetUniqueFormaliteStatusRefused(Provision provision) throws OsirisException;

        public void notifyGuichetUniqueFormaliteSigned(Provision provision) throws OsirisException;

        public Notification addOrUpdatePersonnalNotification(Notification notification);

        public void purgeNotification();

        public List<String> getAllNotificationTypes();

        public List<Notification> getNotificationsForCustomerOrder(Integer customerOrderId) throws OsirisException;

        public List<Notification> getNotificationsForQuotation(Integer quotationId) throws OsirisException;

        public List<Notification> getNotificationsForProvision(Integer provisionId) throws OsirisException;

        public List<Notification> getNotificationsForInvoice(Integer invoiceId) throws OsirisException;

        public List<Notification> getNotificationsForService(Integer serviceId) throws OsirisException;

        public List<Notification> getNotificationsForAffaire(Integer affaireId) throws OsirisException;

        public List<Notification> getNotificationsForTiers(Integer tiersId) throws OsirisException;

        public List<Notification> getNotificationsForResponsable(Integer responsableId) throws OsirisException;

        public void notifyNewCandidacy(Candidacy candidacy) throws OsirisException;

        public void notifyIncidentReportAsked(IncidentReport incident) throws OsirisException;

        public void notifyQuotationModified(CustomerOrder customerOrder);

        public void notifyImmediateAffactationOfOrder(CustomerOrder customerOrder) throws OsirisException;

        public void notifyBodaccNoticeAddToProvision(Provision provision, BodaccNotice notice);

        public void notifyBaloNoticeAddToProvision(Provision provision, BaloNotice notice);

        public void notifyJoNoticeAddToProvision(Provision provision, JoNotice notice);
}
